lib/gruf/controllers/base.rb

```diff
@@ -0,0 +1,103 @@
+# coding: utf-8
+# Copyright (c) 2017-present, BigCommerce Pty. Ltd. All rights reserved
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+# OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+require_relative 'request'
+
+module Gruf
+  module Controllers
+    ##
+    # Base controller object for Gruf gRPC requests
+    #
+    class Base
+      include Gruf::Errors::Handlers
+
+      attr_reader :request
+
+      ##
+      # Initialize the controller within the given request context
+      #
+      # @param [Symbol] method_key
+      # @param [GRPC::GenericService] service
+      # @param [GRPC::RpcDesc] rpc_desc
+      # @param [GRPC::ActiveCall] active_call
+      # @param [Google::Protobuf::MessageExts] message
+      #
+      def initialize(method_key:, service:, rpc_desc:, active_call:, message:)
+        @request = Request.new(
+          method_key: method_key,
+          service: service,
+          rpc_desc: rpc_desc,
+          active_call: active_call,
+          message: message
+        )
+      end
+
+      ##
+      # Bind the controller to the given service and add it to the service registry
+      #
+      # @param [GRPC::GenericService] service
+      #
+      def self.bind(service)
+        self_klass = self
+        Gruf.services << service.name.constantize
+        service.rpc_descs.each do |name, rpc_desc|
+          method_key = name.to_s.underscore.to_sym
+
+          service.class_eval do
+            if rpc_desc.request_response?
```

Comment:
🍺 It would be nice to encapsulate this logic into a `Gruf::Controllers::MethodFactory` (or something like that)